core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -451,7 +411,7 @@ private boolean balance(boolean onlyAssign) {
          * to the nodes we relocated them from.
          */
         private String[] buildWeightOrderedIndices(NodeSorter sorter) {
-            final String[] indices = this.indices.toArray(new String[this.indices.size()]);
+            final String[] indices = allocation.routingTable().indicesRouting().keys().toArray(String.class);
```

Comment:
maybe another optimization is to assign the indices array to a member and then we only do it once? if there are lots of indices it could be costly to recreate that array over and over again?